# hacker news project
on this project we gonna look at the hacker news posts at this dataset 
we're specifically intrested in postes whose titles begin with either Ask HN or Show HN. users submit Ask HN posts to ask the hacker news community a specific question.

likewise, users submit Show Hn posts to show Hacker news community a project, product or just generally something intresting.

we'll compare these two types of posts to determine the following:
* Do Ask Hn or Show HN recieve more comments on avegare?
* Do posts created at certain time recieve more comments on avegare? 


In [1]:
from csv import reader

opened_file = open("hacker_news.csv")
read_file = reader(opened_file)
hn = list(read_file)
print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


now we gonna extract the first row wich is the header and remove it from the dataset variable hn


In [2]:
headers = hn[0]
hn = hn[1:]
print(headers)
print("\n")
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


Now that we've removed the headers from hn, we're ready to filter our data. Since we're only concerned with post titles beginning with Ask HN or Show HN, we'll create new lists of lists containing just the data for those titles.

In [3]:
ask_posts = []
show_posts = []
other_posts = []
for each in hn:
    title = each[1]
    title = title.lower()
    
    if title.startswith("ask hn"):
        ask_posts.append(each)
    elif title.startswith("show hn"):
        show_posts.append(each)
    else:
        other_posts.append(each)

print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


In [4]:
ask_comments = []

for each in ask_posts:
    comments = int(each[4])
    ask_comments.append(comments)

show_comments = []

for each in show_posts:
    comments = int(each[4])
    show_comments.append(comments)

    
avg_ask_com = sum(ask_comments)/len(ask_comments)
avg_show_com = sum(show_comments)/len(show_comments)

print(avg_ask_com)
print("\n")
print(avg_show_com)

14.038417431192661


10.31669535283993


we analize either aks posts or show post get more comments average, we find that post in the ask hn category has a comment average of 14.038417431192661 per post while the show hn category has an average of 10.31669535283993 per post

## analizing posts per post time

Next, we'll determine if ask posts created at a certain time are more likely to attract comments. We'll use the following steps to perform this analysis:

1. Calculate the amount of ask posts created in each hour of the day, along with the number of comments received.
1. Calculate the average number of comments ask posts receive by hour created.

In [5]:
import datetime as dt
results_list = []
for each in ask_posts:
    created_at = each[6]
    num_comm = int(each[4])
    
    result = [created_at,num_comm]
    results_list.append(result)

    
#print(results_list) 
    
counts_by_hour = {} #contains the number of ask posts created during each hour of the day.
comments_by_hour = {} #contains the corresponding number of comments ask posts created at each hour received.

for each in results_list:
    date = each[0]
    date = dt.datetime.strptime(date,"%m/%d/%Y %H:%M")
    post_hour = dt.datetime.strftime(date,"%H")
    
    if post_hour in counts_by_hour:
        counts_by_hour[post_hour] += 1
        comments_by_hour[post_hour] += each[1]
    else:
        counts_by_hour[post_hour] = 1
        comments_by_hour[post_hour] = each[1]
        
        
        
print(counts_by_hour)
print("\n")
print(comments_by_hour)        


{'09': 45, '13': 85, '10': 59, '14': 107, '16': 108, '23': 68, '12': 73, '17': 100, '15': 116, '21': 109, '20': 80, '02': 58, '18': 109, '03': 54, '05': 46, '19': 110, '01': 60, '22': 71, '08': 48, '04': 47, '00': 55, '06': 44, '07': 34, '11': 58}


{'09': 251, '13': 1253, '10': 793, '14': 1416, '16': 1814, '23': 543, '12': 687, '17': 1146, '15': 4477, '21': 1745, '20': 1722, '02': 1381, '18': 1439, '03': 421, '05': 464, '19': 1188, '01': 683, '22': 479, '08': 492, '04': 337, '00': 447, '06': 397, '07': 267, '11': 641}


In [9]:
avg_by_hour = []

for key in comments_by_hour:
    comment = comments_by_hour[key]
    post = counts_by_hour[key]
    hour = key
    avg = [hour,comment/post]
    avg_by_hour.append(avg)
print(avg_by_hour)

[['09', 5.5777777777777775], ['13', 14.741176470588234], ['10', 13.440677966101696], ['14', 13.233644859813085], ['16', 16.796296296296298], ['23', 7.985294117647059], ['12', 9.41095890410959], ['17', 11.46], ['15', 38.5948275862069], ['21', 16.009174311926607], ['20', 21.525], ['02', 23.810344827586206], ['18', 13.20183486238532], ['03', 7.796296296296297], ['05', 10.08695652173913], ['19', 10.8], ['01', 11.383333333333333], ['22', 6.746478873239437], ['08', 10.25], ['04', 7.170212765957447], ['00', 8.127272727272727], ['06', 9.022727272727273], ['07', 7.852941176470588], ['11', 11.051724137931034]]


now we sorted the results

In [10]:
swap_avg_by_hour = []
for each in avg_by_hour:
    swap_list = [each[1],each[0]]
    swap_avg_by_hour.append(swap_list)
    
print(swap_avg_by_hour)

[[5.5777777777777775, '09'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [16.796296296296298, '16'], [7.985294117647059, '23'], [9.41095890410959, '12'], [11.46, '17'], [38.5948275862069, '15'], [16.009174311926607, '21'], [21.525, '20'], [23.810344827586206, '02'], [13.20183486238532, '18'], [7.796296296296297, '03'], [10.08695652173913, '05'], [10.8, '19'], [11.383333333333333, '01'], [6.746478873239437, '22'], [10.25, '08'], [7.170212765957447, '04'], [8.127272727272727, '00'], [9.022727272727273, '06'], [7.852941176470588, '07'], [11.051724137931034, '11']]


In [12]:
sorted_swap = sorted(swap_avg_by_hour,reverse=True)
print(sorted_swap)

[[38.5948275862069, '15'], [23.810344827586206, '02'], [21.525, '20'], [16.796296296296298, '16'], [16.009174311926607, '21'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [13.20183486238532, '18'], [11.46, '17'], [11.383333333333333, '01'], [11.051724137931034, '11'], [10.8, '19'], [10.25, '08'], [10.08695652173913, '05'], [9.41095890410959, '12'], [9.022727272727273, '06'], [8.127272727272727, '00'], [7.985294117647059, '23'], [7.852941176470588, '07'], [7.796296296296297, '03'], [7.170212765957447, '04'], [6.746478873239437, '22'], [5.5777777777777775, '09']]


In [17]:
print("Top 5 Hours for Ask Posts Comments")
for each in sorted_swap[:5]:
    time = each[1]
    time = dt.datetime.strptime(time,"%H" )
    time = dt.datetime.strftime(time,"%H:%M")
    avg = each[0]
    forma = f"{time}: {avg:.2f} average comments per post"
    print(forma)


Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


we find the best time to post with more chances to get comments.  
the time zone that this data set was based on is the Eastern Time in the US, UTC−04:00
converting this time zone for me the best time to post on hacker news is 14:00 local time